src/main/java/graphql/execution/AsynchronousExecutionStrategy.java

```diff
@@ -0,0 +1,223 @@
+        } else if (fieldType instanceof GraphQLScalarType) {
+            return CompletableFuture.completedFuture(completeValueForScalar((GraphQLScalarType) fieldType, result));
+        } else if (fieldType instanceof GraphQLEnumType) {
+            return CompletableFuture.completedFuture(completeValueForEnum((GraphQLEnumType) fieldType, result));
+        }
+
+
+        GraphQLObjectType resolvedType;
+        if (fieldType instanceof GraphQLInterfaceType) {
+            TypeResolutionParameters resolutionParams = TypeResolutionParameters.newParameters()
+                            .graphQLInterfaceType((GraphQLInterfaceType) fieldType)
+                            .field(fields.get(0))
+                            .value(parameters.source())
+                            .argumentValues(parameters.arguments())
+                            .schema(executionContext.getGraphQLSchema()).build();
+            resolvedType = resolveTypeForInterface(resolutionParams);
+
+        } else if (fieldType instanceof GraphQLUnionType) {
+            TypeResolutionParameters resolutionParams = TypeResolutionParameters.newParameters()
+                            .graphQLUnionType((GraphQLUnionType) fieldType)
+                            .field(fields.get(0))
+                            .value(parameters.source())
+                            .argumentValues(parameters.arguments())
+                            .schema(executionContext.getGraphQLSchema()).build();
+            resolvedType = resolveTypeForUnion(resolutionParams);
+        } else {
+            resolvedType = (GraphQLObjectType) fieldType;
+        }
+
+        FieldCollectorParameters collectorParameters = newParameters(executionContext.getGraphQLSchema(), resolvedType)
+                        .fragments(executionContext.getFragmentsByName())
+                        .variables(executionContext.getVariables())
+                        .build();
+
+        Map<String, List<Field>> subFields = fieldCollector.collectFields(collectorParameters, fields);
+
+        ExecutionParameters newParameters = ExecutionParameters.newParameters()
+                        .typeInfo(typeInfo.asType(resolvedType))
+                        .fields(subFields)
+                        .source(result).build();
+
+        // Calling this from the executionContext to ensure we shift back from mutation strategy to the query strategy.
+
+        ExecutionResult executionResult = executionContext.getQueryStrategy().execute(executionContext, newParameters);
+        if(!(executionResult.getData() instanceof CompletionStage)) {
+            return CompletableFuture.completedFuture(executionResult);
+        } else {
+            return ((CompletionStage) executionResult.getData()).handleAsync((resultMap,th) ->
+                            new ExecutionResultImpl(resultMap,executionResult.getErrors())
+            );
+
+        }
+    }
+
```

Comment:
Mostly pondering here.

I wonder if we should make the complete value list truly async.  Should it join back as one result at this point?

Is there any disadvantage is leaving it purely async for many items in a list?  Is there any expectation that is this point you have resolved the value so it can be sent onto future calls?